ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/store/dao/JdbcLedgerDaoCompletionsSpec.scala

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.store.dao
+
+import java.time.Instant
+import java.util.UUID
+
+import akka.stream.scaladsl.Sink
+import com.digitalasset.daml.lf.data.Ref.Party
+import com.digitalasset.ledger.ApplicationId
+import com.digitalasset.ledger.api.domain.RejectionReason
+import com.digitalasset.platform.ApiOffset
+import com.digitalasset.platform.store.{CompletionFromTransaction, PersistenceEntry}
+import com.digitalasset.platform.store.entries.LedgerEntry
+import org.scalatest.{AsyncFlatSpec, Matchers, OptionValues}
+
+import scala.concurrent.Future
+
+private[dao] trait JdbcLedgerDaoCompletionsSpec extends OptionValues {
+  this: AsyncFlatSpec with Matchers with JdbcLedgerDaoSuite =>
+
+  behavior of "JdbcLedgerDao (completions)"
+
+  it should "return the expected completion for an accepted transaction" in {
+    for {
+      from <- ledgerDao.lookupLedgerEnd()
+      (offset, tx) <- storeCreateTransaction()
+      to <- ledgerDao.lookupLedgerEnd()
+      (_, response) <- ledgerDao.completions
+        .getCommandCompletions(from, to, tx.applicationId.get, Set(tx.submittingParty.get))
+        .runWith(Sink.head)
+    } yield {
+      val receivedOffset =
+        ApiOffset.assertFromString(response.checkpoint.value.offset.value.value.absolute.value)
+      receivedOffset shouldBe offset
+
+      response.completions should have length 1
+      val completion = response.completions.head
+
+      completion.transactionId shouldBe tx.transactionId
+      completion.commandId shouldBe tx.commandId.get
+      completion.status.value.code shouldBe io.grpc.Status.Code.OK.value()
+    }
+  }
+
+  private val applicationId: ApplicationId = "JdbcLedgerDaoCompletionsSpec"
+  private val party: Party = "JdbcLedgerDaoCompletionsSpec"
+  private val parties: Set[Party] = Set(party)
+
+  private def rejectWith(reason: RejectionReason): PersistenceEntry.Rejection =
+    PersistenceEntry.Rejection(
+      LedgerEntry.Rejection(
+        recordTime = Instant.now,
+        commandId = UUID.randomUUID().toString,
+        applicationId = applicationId,
+        submitter = party,
+        rejectionReason = reason,
+      )
+    )
+
+  it should "return the expected completion for a rejection" in {
+    val offset = nextOffset()
+    val rejection = rejectWith(RejectionReason.Inconsistent(""))
+    for {
+      from <- ledgerDao.lookupLedgerEnd()
+      _ <- ledgerDao.storeLedgerEntry(offset, rejection)
+      to <- ledgerDao.lookupLedgerEnd()
+      (_, response) <- ledgerDao.completions
+        .getCommandCompletions(from, to, applicationId, parties)
+        .runWith(Sink.head)
+    } yield {
+      val receivedOffset =
+        ApiOffset.assertFromString(response.checkpoint.value.offset.value.value.absolute.value)
+      receivedOffset shouldBe offset
+
+      response.completions should have length 1
+      val completion = response.completions.head
```

Comment:
```suggestion
      val completion = response.completions.loneElement
```

Reply:
For future reference, this is added by the trait `LoneElement`.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5072/commits/3785e7c393670e0a218c4680b173a8d3ad34cbf2